// @fluentui-contrib/react-virtualizer/lib-commonjs/components/VirtualizerScrollViewDynamic/useVirtualizerScrollViewDynamic.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "useVirtualizerScrollViewDynamic_unstable", {
    enumerable: true,
    get: function() {
        return useVirtualizerScrollViewDynamic_unstable;
    }
});
const _interop_require_wildcard = require("@swc/helpers/_/_interop_require_wildcard");
const _react = /*#__PURE__*/ _interop_require_wildcard._(require("react"));
const _reactutilities = require("@fluentui/react-utilities");
const _useVirtualizer = require("../Virtualizer/useVirtualizer");
const _Hooks = require("../../Hooks");
const _Utilities = require("../../Utilities");
const _useMeasureList = require("../../hooks/useMeasureList");
const _useDynamicPagination = require("../../hooks/useDynamicPagination");
function useVirtualizerScrollViewDynamic_unstable(props) {
    'use no memo';
    var _imperativeVirtualizerRef_current;
    const contextState = (0, _Utilities.useVirtualizerContextState_unstable)(props.virtualizerContext);
    const { imperativeRef, axis = 'vertical', reversed, imperativeVirtualizerRef, enablePagination = false, bufferItems: _bufferItems, bufferSize: _bufferSize, enableScrollAnchor, gap = 0 } = props;
    const sizeTrackingArray = _react.useRef(new Array(props.numItems).fill(props.itemSize));
    const getChildSizeAuto = (index)=>{
        if (sizeTrackingArray.current.length <= index || sizeTrackingArray.current[index] <= 0) {
            // Default size for initial state or untracked
            return props.itemSize;
        }
        /* Required to be defined prior to our measure function
     * we use a sizing array ref that we will update post-render
     */ return sizeTrackingArray.current[index];
    };
    var _props_axis, _props_getItemSize;
    const { virtualizerLength, bufferItems, bufferSize, scrollRef, containerSizeRef, updateScrollPosition } = (0, _Hooks.useDynamicVirtualizerMeasure)({
        defaultItemSize: props.itemSize,
        direction: (_props_axis = props.axis) != null ? _props_axis : 'vertical',
        getItemSize: (_props_getItemSize = props.getItemSize) != null ? _props_getItemSize : getChildSizeAuto,
        virtualizerContext: contextState,
        numItems: props.numItems,
        bufferItems: _bufferItems,
        bufferSize: _bufferSize,
        gap
    });
    const _imperativeVirtualizerRef = (0, _reactutilities.useMergedRefs)(_react.useRef(null), imperativeVirtualizerRef);
    var _contextState_contextIndex;
    const paginationRef = (0, _useDynamicPagination.useDynamicVirtualizerPagination)({
        axis,
        progressiveItemSizes: (_imperativeVirtualizerRef_current = _imperativeVirtualizerRef.current) == null ? void 0 : _imperativeVirtualizerRef_current.progressiveSizes,
        actualNodeSizes: sizeTrackingArray,
        virtualizerLength,
        currentIndex: (_contextState_contextIndex = contextState == null ? void 0 : contextState.contextIndex) != null ? _contextState_contextIndex : 0
    }, enablePagination);
    // Store the virtualizer length as a ref for imperative ref access
    const virtualizerLengthRef = _react.useRef(virtualizerLength);
    if (virtualizerLengthRef.current !== virtualizerLength) {
        virtualizerLengthRef.current = virtualizerLength;
    }
    const localScrollRef = _react.useRef(null);
    const scrollViewRef = (0, _reactutilities.useMergedRefs)(props.scrollViewRef, scrollRef, paginationRef, localScrollRef);
    const scrollCallbackRef = _react.useRef(null);
    _react.useImperativeHandle(imperativeRef, ()=>{
        var _imperativeVirtualizerRef_current;
        return {
            scrollToPosition (position, behavior = 'auto', index, callback) {
                if (callback) {
                    scrollCallbackRef.current = callback != null ? callback : null;
                }
                if (_imperativeVirtualizerRef.current) {
                    var _scrollViewRef_current;
                    if (index !== undefined) {
                        _imperativeVirtualizerRef.current.setFlaggedIndex(index);
                    }
                    const positionOptions = axis == 'vertical' ? {
                        top: position
                    } : {
                        left: position
                    };
                    (_scrollViewRef_current = scrollViewRef.current) == null ? void 0 : _scrollViewRef_current.scrollTo({
                        behavior,
                        ...positionOptions
                    });
                }
            },
            scrollTo (index, behavior = 'auto', callback) {
                scrollCallbackRef.current = callback != null ? callback : null;
                if (_imperativeVirtualizerRef.current) {
                    const progressiveSizes = _imperativeVirtualizerRef.current.progressiveSizes.current;
                    const totalSize = progressiveSizes && (progressiveSizes == null ? void 0 : progressiveSizes.length) > 0 ? progressiveSizes[Math.max(progressiveSizes.length - 1, 0)] : 0;
                    _imperativeVirtualizerRef.current.setFlaggedIndex(index);
                    var _props_getItemSize;
                    (0, _Utilities.scrollToItemDynamic)({
                        index,
                        getItemSize: (_props_getItemSize = props.getItemSize) != null ? _props_getItemSize : getChildSizeAuto,
                        totalSize,
                        scrollViewRef: scrollViewRef,
                        axis,
                        reversed,
                        behavior,
                        gap
                    });
                }
            },
            currentIndex: (_imperativeVirtualizerRef_current = _imperativeVirtualizerRef.current) == null ? void 0 : _imperativeVirtualizerRef_current.currentIndex,
            virtualizerLength: virtualizerLengthRef,
            sizeTrackingArray
        };
    }, [
        axis,
        scrollViewRef,
        reversed,
        _imperativeVirtualizerRef
    ]);
    const handleRenderedIndex = (index)=>{
        if (scrollCallbackRef.current) {
            scrollCallbackRef.current(index);
        }
    };
    var _props_getItemSize1;
    const virtualizerState = (0, _useVirtualizer.useVirtualizer_unstable)({
        ...props,
        getItemSize: (_props_getItemSize1 = props.getItemSize) != null ? _props_getItemSize1 : getChildSizeAuto,
        virtualizerLength,
        bufferItems,
        bufferSize,
        virtualizerContext: contextState,
        imperativeVirtualizerRef: _imperativeVirtualizerRef,
        onRenderedFlaggedIndex: handleRenderedIndex,
        containerSizeRef,
        scrollViewRef,
        updateScrollPosition
    });
    const requestScrollBy = _react.useCallback((sizeChange)=>{
        // Handle any size changes so that scroll view doesn't jump around
        if (enableScrollAnchor) {
            var _localScrollRef_current;
            (_localScrollRef_current = localScrollRef.current) == null ? void 0 : _localScrollRef_current.scrollBy({
                top: axis === 'vertical' ? sizeChange : 0,
                left: axis === 'vertical' ? 0 : sizeChange,
                behavior: 'instant'
            });
        }
    }, [
        enableScrollAnchor,
        axis,
        localScrollRef
    ]);
    const measureObject = (0, _useMeasureList.useMeasureList)({
        currentIndex: Math.max(virtualizerState.virtualizerStartIndex, 0),
        totalLength: props.numItems,
        defaultItemSize: props.itemSize,
        sizeTrackingArray,
        axis,
        virtualizerLength,
        requestScrollBy
    });
    // Enables auto-measuring and tracking post render sizes externally
    _react.Children.map(virtualizerState.virtualizedChildren, (child, index)=>{
        if (/*#__PURE__*/ _react.isValidElement(child)) {
            virtualizerState.virtualizedChildren[index] = /*#__PURE__*/ _react.createElement(child.type, {
                ...child.props,
                key: child.key,
                ref: (element)=>{
                    if (Object.prototype.hasOwnProperty.call(child, 'ref')) {
                        // We must access this from the child directly, not props (forward ref).
                        // eslint-disable-next-line  @typescript-eslint/no-explicit-any
                        const localRef = child == null ? void 0 : child.ref;
                        if (typeof localRef === 'function') {
                            localRef(element);
                        } else if (localRef) {
                            localRef.current = element;
                        }
                    }
                    // Call the auto-measure ref attachment.
                    measureObject.createIndexedRef(index)(element);
                }
            });
        }
    });
    return {
        ...virtualizerState,
        enableScrollAnchor,
        components: {
            ...virtualizerState.components,
            container: 'div'
        },
        container: _reactutilities.slot.always(props.container, {
            defaultProps: {
                ref: scrollViewRef
            },
            elementType: 'div'
        })
    };
}
