// @babylonjs/inspector/lib/index.js
export { A as Accordion, d as AccordionSection, p as AttachDebugLayer, aI as BooleanBadgePropertyLine, t as BoundProperty, y as BuiltInsExtensionFeed, B as Button, e as ButtonLine, ac as Checkbox, C as CheckboxPropertyLine, aC as ChildWindow, ad as Collapse, an as Color3GradientComponent, aF as Color3GradientList, aJ as Color3PropertyLine, ao as Color4GradientComponent, aG as Color4GradientList, aK as Color4PropertyLine, ae as ColorPickerPopup, ap as ColorStepGradientComponent, ah as ComboBox, a7 as ConstructorFactory, o as ConvertOptions, D as DebugServiceIdentity, q as DetachDebugLayer, ai as DraggableLine, aj as Dropdown, E as ExtensibleAccordion, am as FactorGradientComponent, aE as FactorGradientList, aD as FileUploadLine, a3 as GetPropertyDescriptor, aL as HexPropertyLine, aq as InfoLabel, af as InputHexField, ag as InputHsvField, I as Inspector, a2 as InterceptFunction, a5 as InterceptProperty, a4 as IsPropertyReadonly, aO as LineContainer, L as Link, aM as LinkPropertyLine, w as LinkToEntityPropertyLine, as as List, a1 as MakeDialogTeachingMoment, ar as MakeLazyComponent, h as MakePopoverTeachingMoment, J as MakePropertyHook, a0 as MakeTeachingMoment, M as MessageBar, ak as NumberDropdown, r as NumberDropdownPropertyLine, N as NumberInputPropertyLine, a6 as ObservableCollection, aH as Pane, aP as PlaceholderPropertyLine, at as PositionedPopover, k as PropertiesServiceIdentity, v as Property, aN as PropertyLine, aW as QuaternionPropertyLine, aV as RotationVectorPropertyLine, b as SceneContextIdentity, l as SceneExplorerServiceIdentity, au as SearchBar, av as SearchBox, a9 as SelectionServiceDefinition, a8 as SelectionServiceIdentity, aa as SettingsContextIdentity, m as SettingsServiceIdentity, a as ShellServiceIdentity, ab as ShowInspector, j as SidePaneContainer, aw as SpinButton, S as SpinButtonPropertyLine, n as StatsServiceIdentity, al as StringDropdown, s as StringDropdownPropertyLine, aQ as StringifiedPropertyLine, ax as Switch, aR as SwitchPropertyLine, ay as SyncedSliderInput, aS as SyncedSliderPropertyLine, i as TeachingMoment, aT as TextAreaPropertyLine, aA as TextInput, T as TextInputPropertyLine, aU as TextPropertyLine, az as Textarea, x as Theme, aB as ToggleButton, f as ToolsServiceIdentity, aX as Vector2PropertyLine, V as Vector3PropertyLine, aY as Vector4PropertyLine, $ as useAngleConverters, X as useAsyncResource, F as useColor3Property, G as useColor4Property, Y as useCompactMode, Z as useDisableCopy, O as useEventfulState, K as useInterceptObservable, Q as useObservableCollection, c as useObservableState, R as useOrderedObservableCollection, U as usePollingObservable, u as useProperty, H as useQuaternionProperty, W as useResource, _ as useSidePaneDockOverrides, z as useVector3Property } from './index-ByfjmUIP.js';
import 'react/jsx-runtime';
import 'react';
import '@fluentui/react-components';
import '@fluentui/react-icons';
import '@babylonjs/core/Maths/math.color.js';
import '@babylonjs/core/Maths/math.vector.js';
import '@babylonjs/core/Misc/observable.js';
import '@babylonjs/core/Misc/typeStore.js';
import '@fluentui/react-motion-components-preview';
import 'usehooks-ts';
import '@babylonjs/core/Misc/asyncLock.js';
import '@babylonjs/core/Misc/deferred.js';
import '@babylonjs/core/Misc/logger.js';
import '@babylonjs/core/Maths/math.scalar.functions.js';
import '@fluentui-contrib/react-virtualizer';
import '@babylonjs/core/Misc/uniqueIdGenerator.js';
import '@babylonjs/addons/msdfText/fontAsset.js';
import '@babylonjs/addons/msdfText/textRenderer.js';
import '@babylonjs/core/Debug/physicsViewer.js';
import '@babylonjs/core/Materials/Textures/texture.js';
import '@babylonjs/core/Materials/materialFlags.js';
import '@babylonjs/core/Materials/standardMaterial.js';
import '@babylonjs/core/Meshes/Builders/groundBuilder.js';
import '@babylonjs/core/Misc/tools.js';
import '@babylonjs/core/Rendering/utilityLayerRenderer.js';
import '@babylonjs/materials/grid/gridMaterial.js';
import '@babylonjs/core/Misc/dataStorage.js';
import '@babylonjs/core/Instrumentation/engineInstrumentation.js';
import '@babylonjs/core/Instrumentation/sceneInstrumentation.js';
import '@babylonjs/core/Engines/AbstractEngine/abstractEngine.timeQuery.js';
import '@babylonjs/core/Engines/Extensions/engine.query.js';
import '@babylonjs/core/Engines/WebGPU/Extensions/engine.query.js';
import '@babylonjs/core/Misc/PerformanceViewer/performanceViewerCollectionStrategies.js';
import '@babylonjs/core/Misc/PerformanceViewer/performanceViewerSceneExtension.js';
import '@babylonjs/core/Misc/pressureObserverWrapper.js';
import '@babylonjs/core/Maths/math.scalar.js';
import '@babylonjs/core/Misc/PerformanceViewer/performanceViewerCollector.js';
import '@babylonjs/core/Engines/abstractEngine.js';
import 'react-dom/client';
import '@babylonjs/core/FrameGraph/frameGraphUtils.js';
import '@babylonjs/core/Gizmos/cameraGizmo.js';
import '@babylonjs/core/Gizmos/lightGizmo.js';
import '@babylonjs/core/Bones/bone.js';
import '@babylonjs/core/Cameras/camera.js';
import '@babylonjs/core/Gizmos/gizmoManager.js';
import '@babylonjs/core/Lights/light.js';
import '@babylonjs/core/Meshes/abstractMesh.js';
import '@babylonjs/core/node.js';
import '@babylonjs/core/Animations/animationGroup.js';
import '@babylonjs/core/Animations/animation.js';
import '@babylonjs/core/Animations/animationPropertiesOverride.js';
import '@babylonjs/core/Audio/sound.js';
import '@babylonjs/core/Cameras/arcRotateCamera.js';
import '@babylonjs/core/Cameras/followCamera.js';
import '@babylonjs/core/Cameras/freeCamera.js';
import '@babylonjs/core/Cameras/geospatialCamera.js';
import '@babylonjs/core/Cameras/targetCamera.js';
import '@babylonjs/core/scene.js';
import '@babylonjs/core/FrameGraph/frameGraph.js';
import '@babylonjs/core/Lights/directionalLight.js';
import '@babylonjs/core/Lights/hemisphericLight.js';
import '@babylonjs/core/Lights/pointLight.js';
import '@babylonjs/core/Lights/rectAreaLight.js';
import '@babylonjs/core/Lights/shadowLight.js';
import '@babylonjs/core/Lights/spotLight.js';
import '@babylonjs/core/Lights/Shadows/cascadedShadowGenerator.js';
import '@babylonjs/core/Debug/directionalLightFrustumViewer.js';
import '@babylonjs/core/Lights/Shadows/shadowGenerator.js';
import '@babylonjs/core/Lights/Shadows/shadowGeneratorSceneComponent.js';
import '@babylonjs/core/Materials/material.js';
import '@babylonjs/core/Materials/multiMaterial.js';
import '@babylonjs/core/Materials/Node/nodeMaterial.js';
import '@babylonjs/core/Materials/PBR/openpbrMaterial.js';
import '@babylonjs/core/Materials/PBR/pbrBaseMaterial.js';
import '@babylonjs/materials/sky/skyMaterial.js';
import '@babylonjs/core/Engines/constants.js';
import '@babylonjs/core/Engines/engine.js';
import '@babylonjs/core/Particles/particleSystem.js';
import '@babylonjs/core/Materials/Node/Blocks/gradientBlock.js';
import '@babylonjs/core/Materials/Node/Enums/nodeMaterialBlockConnectionPointTypes.js';
import '@babylonjs/core/Misc/gradients.js';
import '@babylonjs/core/Misc/fileTools.js';
import '@babylonjs/core/Materials/Textures/cubeTexture.js';
import '@babylonjs/core/Meshes/GaussianSplatting/gaussianSplattingMesh.js';
import '@babylonjs/core/Meshes/mesh.js';
import '@babylonjs/core/Debug/skeletonViewer.js';
import '@babylonjs/core/Meshes/buffer.js';
import '@babylonjs/core/Meshes/Builders/linesBuilder.js';
import '@babylonjs/core/Meshes/instancedMesh.js';
import '@babylonjs/core/Rendering/renderingManager.js';
import '@babylonjs/core/Rendering/edgesRenderer.js';
import '@babylonjs/core/Rendering/outlineRenderer.js';
import '@babylonjs/core/Particles/gpuParticleSystem.js';
import '@babylonjs/core/Particles/Node/nodeParticleSystemSet.helper.js';
import '@babylonjs/core/Particles/particleHelper.js';
import '@babylonjs/core/Particles/EmitterTypes/boxParticleEmitter.js';
import '@babylonjs/core/Particles/EmitterTypes/coneParticleEmitter.js';
import '@babylonjs/core/Particles/EmitterTypes/cylinderParticleEmitter.js';
import '@babylonjs/core/Particles/EmitterTypes/hemisphericParticleEmitter.js';
import '@babylonjs/core/Particles/EmitterTypes/meshParticleEmitter.js';
import '@babylonjs/core/Particles/EmitterTypes/pointParticleEmitter.js';
import '@babylonjs/core/Particles/EmitterTypes/sphereParticleEmitter.js';
import '@babylonjs/core/Particles/attractor.js';
import '@babylonjs/core/Meshes/Builders/sphereBuilder.js';
import '@babylonjs/core/Particles/Node/Blocks/particleInputBlock.js';
import '@babylonjs/core/Particles/Node/Blocks/Update/updateAttractorBlock.js';
import '@babylonjs/core/Particles/Node/Enums/nodeParticleBlockConnectionPointTypes.js';
import '@babylonjs/core/Meshes/transformNode.js';
import '@babylonjs/core/Physics/v2/IPhysicsEnginePlugin.js';
import '@babylonjs/core/Physics/v2/physicsEngineComponent.js';
import '@babylonjs/core/PostProcesses/postProcess.js';
import '@babylonjs/core/PostProcesses/RenderPipeline/postProcessRenderPipeline.js';
import '@babylonjs/core/PostProcesses/RenderPipeline/Pipelines/defaultRenderingPipeline.js';
import '@babylonjs/core/PostProcesses/RenderPipeline/Pipelines/lensRenderingPipeline.js';
import '@babylonjs/core/PostProcesses/RenderPipeline/Pipelines/ssaoRenderingPipeline.js';
import '@babylonjs/core/PostProcesses/RenderPipeline/Pipelines/ssao2RenderingPipeline.js';
import '@babylonjs/core/PostProcesses/RenderPipeline/Pipelines/ssrRenderingPipeline.js';
import '@babylonjs/core/Rendering/IBLShadows/iblShadowsRenderPipeline.js';
import '@babylonjs/core/Materials/imageProcessingConfiguration.js';
import '@babylonjs/core/Bones/skeleton.js';
import '@babylonjs/core/Sprites/sprite.js';
import '@babylonjs/core/Sprites/spriteManager.js';
import '@babylonjs/core/Misc/textureTools.js';
import '@babylonjs/core/Materials/Textures/baseTexture.js';
import '@babylonjs/core/Materials/Textures/multiRenderTarget.js';
import '@babylonjs/core/Materials/Textures/renderTargetTexture.js';
import '@babylonjs/core/Materials/Textures/thinTexture.js';
import '@babylonjs/core/Events/keyboardEvents.js';
import '@babylonjs/core/Events/pointerEvents.js';
import '@babylonjs/core/Materials/Textures/htmlElementTexture.js';
import '@babylonjs/core/Materials/shaderMaterial.js';
import '@babylonjs/core/Meshes/Builders/planeBuilder.js';
import '@babylonjs/core/Lights/Clustered/clusteredLightContainer.js';
import '@babylonjs/core/Rendering/boundingBoxRenderer.js';
import '@babylonjs/core/PostProcesses/RenderPipeline/postProcessRenderPipelineManagerSceneComponent.js';
import '@babylonjs/core/Sprites/spriteSceneComponent.js';
import '@babylonjs/core/Materials/Textures/dynamicTexture.js';
import '@babylonjs/core/Misc/equirectangularCapture.js';
import '@babylonjs/core/Misc/sceneRecorder.js';
import '@babylonjs/core/Misc/screenshotTools.js';
import '@babylonjs/core/Misc/videoRecorder.js';
import '@babylonjs/core/Misc/sceneSerializer.js';
import '@babylonjs/core/Misc/environmentTextureTools.js';
import '@babylonjs/core/Loading/sceneLoader.js';
import '@babylonjs/core/Misc/filesInput.js';
import '@babylonjs/loaders/glTF/glTFFileLoader.js';
import '@babylonjs/loaders/glTF/glTFValidation.js';
import '@babylonjs/core/Engines/engineStore.js';
import '@babylonjs/core/Debug/debugLayer.js';
import '@babylonjs/core/Misc/lazy.js';
